### PR TITLE
added referral logic

### PR DESCRIPTION
**Pull request by FidelVe**
the following PR adds the required logic for the implementation of the referral feature as discussed.

```
# Simple version

- Main user: Bob
- Friend: Alice

User Bob should be able to share a referral code link. This would look something like this:

**[rewards.hanawallet.io?ref=7fa3ac146645](http://rewards.hanawallet.io/?ref=7fa3ac146645)**

When user Bob generates a sign up through his referral link, which is his friend Alice, two things should happen:

**Alice should receive an extra 2000XP rewards IF she reaches a minimum of 5000XP through task completion**

**Bob should receive an extra 2000XP rewards IF Alice reaches a minimum of 5000XP through task completion**
```

The amount to be earned for being a referrer or a referred can be easily customized as well as the XP threshold values required to earn the task XP, currently those values are defined as:

* for a referred (using a referral code generated from another user): the user will earn 100 XP after the user (this user, the one using the referral code) generates 100 XP
* for a referee (another user using your referral code): the user will earn 150 XP after the OTHER user (the one using your referral code) generates 150 XP

### CODE REVIEW

**Comment by FidelVe** [comment]
when a referral has been processed, the value of the param `isProcessed` gets set up as `true`

```
[
  {
    _id: ObjectId('6748fe323d711feb3e40fe2b'),
    referrerUserAddress: 'hx9bf5c772bdd768a5102625c35808d254eb3434ea',
    referredUserAddress: 'hx1b99d10aa2f9a11ac590fe7102075f3da1e5f4f5',
    referralCode: '294b35b67cbc',
    referrerUserId: ObjectId('6748fe063d711feb3e40fe0e'),
    referredUserId: ObjectId('6748fe323d711feb3e40fe28'),
    isProcessed: true,
    createdAt: ISODate('2024-11-28T23:35:14.118Z'),
    updatedAt: ISODate('2024-11-28T23:35:14.118Z'),
    __v: 0
  }
]
```

File: db-manager/tasks/genericTask.js [hunk 1]
```diff
@@ -228,6 +244,238 @@ async function genericTask(taskInput, db, seedId, callback) {
...
+      // to know the amount of XP required to have before
+      // this task is done
+      let allCriteriaMet = true;
+      for (const eachCriteria of targetTask.criteria) {
+        const conditionFormula = new Function(...eachCriteria.conditionFormula);
+        const criteriaMet = conditionFormula(userXpTotal);
+        if (criteriaMet === false) {
+          allCriteriaMet = false;
+          break;
+        }
```
Comment (R0bi7) [inline db-manager/tasks/genericTask.js]: You could make this a single liner using Array every function (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every)

File: db-manager/tasks/genericTask.js [hunk 2]
```diff
@@ -228,6 +244,238 @@ async function genericTask(taskInput, db, seedId, callback) {
...
+      // to know the amount of XP required to have before
+      // this task is done
+      let allCriteriaMet = true;
+      for (const eachCriteria of targetTask.criteria) {
+        const conditionFormula = new Function(...eachCriteria.conditionFormula);
+        const criteriaMet = conditionFormula(referredUserXpTotal);
+        if (criteriaMet === false) {
+          allCriteriaMet = false;
+          break;
+        }
```
Comment (R0bi7) [inline db-manager/tasks/genericTask.js]: Same as above

File: nestjs-rest-server/src/referral/referral.controller.ts [hunk 3]
```diff
@@ -48,7 +52,12 @@ export class ReferralController {
   })
   async createUserReferral(@Query("referralCode") referralCode: string, @UserAddress() address: string): Promise<void> {
     try {
-      return await this.referralService.createUserReferral(referralCode, address);
+      const referredUser = await this.userDb.getUserByAddress(address);
+
+      if (!referredUser) {
+        throw new InternalServerErrorException("User not found");
```
Comment (R0bi7) [inline nestjs-rest-server/src/referral/referral.controller.ts]: Return Not found server exception (404)

File: nestjs-rest-server/src/referral/referral.controller.ts [hunk 4]
```diff
@@ -48,7 +52,12 @@ export class ReferralController {
...
+      }
+      return await this.referralService.createUserReferral(referralCode, address, referredUser._id);
```
Comment (R0bi7) [inline nestjs-rest-server/src/referral/referral.controller.ts]: I would move logic to the service as controller is merely receiving and forwarding work to service which does the business logic

File: nestjs-rest-server/src/user/user.service.ts [hunk 5]
```diff
@@ -235,23 +235,32 @@ export class UserService {
...
+      // for now we will just continue and the user will not earn
+      // the referral bonus
+      if (referralCode) {
+        try {
+          await this.referralService.createUserReferral(referralCode, publicAddress, rawUser._id);
+        } catch {
+          throw new InternalServerErrorException("Failed to create referral");
+        }
+      }
+      return formatUser(rawUser);
```
Comment (R0bi7) [inline nestjs-rest-server/src/user/user.service.ts]: If we allow creation of an user whilst referral fails, we should gracefully return request - meaning no error throw in line 260. Thus I suggest we just log an exception and return user.

**Review by R0bi7 — CHANGES_REQUESTED** [review]
Looks good overall, small change request

File: nestjs-rest-server/src/user/user.service.ts (hunk 5 above)
Comment (FidelVe) [reply nestjs-rest-server/src/user/user.service.ts]: done

**Review by R0bi7 — APPROVED** [review]

**Pull request merged**